libcst/metadata/scope_provider.py

```diff
@@ -0,0 +1,309 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+
+import abc
+import builtins
+from collections import defaultdict
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Dict, Iterator, List, MutableMapping, Optional, Tuple, Type, Union
+
+import libcst as cst
+from libcst.metadata.base_provider import BatchableMetadataProvider
+from libcst.metadata.expression_context_provider import (
+    ExpressionContext,
+    ExpressionContextProvider,
+)
+
+
+@dataclass(frozen=True)
+class Access:
+    node: cst.Name
+    scope: "Scope"
+
+
+class BaseAssignment:
+    name: str
+    scope: "Scope"
+    __accesses: List[Access]
+
+    def __init__(self, name: str, scope: "Scope") -> None:
+        self.name = name
+        self.scope = scope
+        self.__accesses = []
+
+    def record_access(self, access: Access) -> None:
+        self.__accesses.append(access)
+
+    @property
+    def accesses(self) -> Tuple[Access, ...]:
+        # we don't want to publicly expose the mutable version of this
+        return tuple(self.__accesses)
+
+
+class Assignment(BaseAssignment):
+    node: cst.CSTNode
+
+    def __init__(self, name: str, scope: "Scope", node: cst.CSTNode) -> None:
+        self.node = node
+        super().__init__(name, scope)
+
+
+class BuiltinAssignmemt(BaseAssignment):
+    node: None = None
+
+
+class Scope(abc.ABC):
+    parent: "Scope"
+    globals: "GlobalScope"
+    _assignments: MutableMapping[str, List[BaseAssignment]]
+
+    def __init__(self, parent: "Scope") -> None:
+        super().__init__()
+        self.parent = parent
+        self.globals = parent.globals
+        self._assignments = defaultdict(list)
+
+    def record_assignment(self, name: str, node: cst.CSTNode) -> None:
+        self._assignments[name].append(Assignment(name=name, scope=self, node=node))
+
+    def _contains_in_parent_scope(self, name: str) -> bool:
```

Comment:
Looks like nothing is calling `_contains_in_parent_scope` right now.

I think I implemented this in the AST version because it let me implement `__contains__` more efficiently (no extra tuple construction). If you plan to carry over that small optimization, then leave this here, otherwise remove it.

Reply:
Nice catch. I prefer simple implementation to make the code easier to understand. E.g. I reuse `__getitem__` to implement `__contains__`. Even though it may be slower, we can optimize it later when we met the performance issue. Remove this function for now.